### ISSUE TITLE
This plugin does not seem to work

### ISSUE THREAD

**Issue by fusion809** [ptrv/company-lua#9]
Hi,

I've installed this plugin with M-x package-install and for some reason in Spacemacs and plain old Emacs no auto-completions are provided. Typing in `math.p` doesn't offer any completions, let alone the one I'm hoping for `math.pi`. Same with `math.si`, doesn't lead to any completions let alone the one I want, `math.sin`.  

Thanks for your time,
Brenton

**Comment by ptrv** [comment]
You need to configure the completion backend in order to work, i.e. you need to tell company to use the backend. Have you read the README?

For example:

```emacs-lisp
(defun my-lua-mode-company-init ()
  (setq-local company-backends '((company-lua
                                  company-etags
                                  company-dabbrev-code
                                  company-yasnippet))))
(add-hook 'lua-mode-hook #'my-lua-mode-company-init)
```

**Comment by fusion809** [comment]
I ignored it because I was fairly confident I didn't want company-yasnippet. Snippets don't really interest me.

**Comment by fusion809** [comment]
Just tried adding that to my ~/.spacemacs file and still I don't get the auto-completes. Granted that's not surprising given that I don't have any of the packages installed aside from company-lua. That is because I can't seem to install these packages. company-etags, company-yasnippet, company-dabbrev-code they're all not listed by M-x list-packages. Likewise running M-x package-install and entering their names does not install them. Removing every line mentioning a package from this snippet extra company-lua also doesn't cause the auto-completes to occur.

**Comment by ptrv** [comment]
company-etags, company-yasnippet, company-dabbrev-code are stock company mode completers, no need to install them. 

Do you have `company-mode` enabled?

Actually this would be also enough instead of the code snippet from my post before:
```emacs-lisp
(add-to-list 'company-backends 'company-lua)
```

**Comment by ptrv** [comment]
I'm not a spacemacs user, so please ask also in the spacemacs community how to install and configure a new company-mode completion backend.

**Comment by fusion809** [comment]
Managed to use company-tern without a problem, including in spacemacs, so I doubt it's company that's the main problem. I have this line in my ~/.spacemacs file:

```lisp
 (global-company-mode t)
```

I think that activates it right?

**Comment by fusion809** [comment]
Just tried in vanilla Emacs (without Spacemacs) and had the same problem.

**Comment by ptrv** [comment]
Do you have lua installed?

Can you post the configuration you used in vanilla Emacs to set up the plugin?

**Comment by fusion809** [comment]
Sure do; here is the `~/.emacs`:

```lisp
;; Packages

(package-initialize)

(require 'package)
(setq use-package-always-ensure t)
(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(ansi-color-faces-vector
   [default bold shadow italic underline bold bold-italic bold])
 '(ansi-color-names-vector
   ["%272822" "%F92672" "%A6E22E" "%E6DB74" "%66D9EF" "%FD5FF0" "%A1EFE4" "%F8F8F2"])
 '(ansi-term-color-vector
   [unspecified "%FFFFFF" "%d15120" "%5f9411" "%d2ad00" "%6b82a7" "%a66bab" "%6b82a7" "%505050"] t)
 '(compilation-message-face (quote default))
 '(custom-safe-themes
   (quote
    ("c7a9a68bd07e38620a5508fef62ec079d274475c8f92d75ed0c33c45fbe306bc" "f81a9aabc6a70441e4a742dfd6d10b2bae1088830dc7aba9c9922f4b1bd2ba50" "1dffeecd1565d04cd2059234e872cd80fcbe813488602d5c42b5c9e576924d9f" "a2dd771a05705be2a6e6adb6ddbc7a27ebf49edab1dffdbefe243096becba7c9" "c1390663960169cd92f58aad44ba3253227d8f715c026438303c09b9fb66cdfb" "28c07d8e75ed7e2ea9154239caa4f9d3fc831cf2340ea3e307ee49b4321dc7f5" "98cc377af705c0f2133bb6d340bf0becd08944a588804ee655809da5d8140de6" "5dc0ae2d193460de979a463b907b4b2c6d2c9c4657b2e9e66b8898d2592e3de5" "74c328628e9ced700633c9740666e56f40da5062d93b99b7f2cabb54d0b9194a" "a800120841da457aa2f86b98fb9fd8df8ba682cebde033d7dbf8077c1b7d677a" default)))
 '(elfeed-feeds
   (quote
    ("https://github.com/Microsoft/vscode/releases.atom" "https://github.com/atom/atom/releases.atom")))
 '(ensime-sem-high-faces
   (quote
    ((var :foreground "%000000" :underline
	  (:style wave :color "yellow"))
     (val :foreground "%000000")
     (varField :foreground "%600e7a" :slant italic)
     (valField :foreground "%600e7a" :slant italic)
     (functionCall :foreground "%000000" :slant italic)
     (implicitConversion :underline
			 (:color "%c0c0c0"))
     (implicitParams :underline
		     (:color "%c0c0c0"))
     (operator :foreground "%000080")
     (param :foreground "%000000")
     (class :foreground "%20999d")
     (trait :foreground "%20999d" :slant italic)
     (object :foreground "%5974ab" :slant italic)
     (package :foreground "%000000")
     (deprecated :strike-through "%000000"))))
 '(fci-rule-character-color "%d9d9d9")
 '(fci-rule-color "%20240E")
 '(highlight-changes-colors (quote ("%FD5FF0" "%AE81FF")))
 '(highlight-tail-colors
   (quote
    (("%20240E" . 0)
     ("%679A01" . 20)
     ("%4BBEAE" . 30)
     ("%1DB4D0" . 50)
     ("%9A8F21" . 60)
     ("%A75B00" . 70)
     ("%F309DF" . 85)
     ("%20240E" . 100))))
 '(hl-sexp-background-color "%1c1f26")
 '(magit-diff-use-overlays nil)
 '(package-selected-packages
   (quote
    (company-lua company auto-complete-clang auto-complete less-css-mode dired-icon all-the-icons-dired gnuplot-mode white-theme vala-mode twilight-bright-theme tabbar-ruler sage-shell-mode rust-mode rpm-spec-mode pyvenv python-pylint python-cell pkgbuild-mode markdown-mode lua-mode highlight-indentation haskell-mode elfeed dockerfile-mode clojure-mode)))
 '(pos-tip-background-color "%A6E22E")
 '(pos-tip-foreground-color "%272822")
 '(speedbar-show-unknown-files t)
 '(vc-annotate-background nil)
 '(vc-annotate-color-map
   (quote
    ((20 . "%F92672")
     (40 . "%CF4F1F")
     (60 . "%C26C0F")
     (80 . "%E6DB74")
     (100 . "%AB8C00")
     (120 . "%A18F00")
     (140 . "%989200")
     (160 . "%8E9500")
     (180 . "%A6E22E")
     (200 . "%729A1E")
     (220 . "%609C3C")
     (240 . "%4E9D5B")
     (260 . "%3C9F79")
     (280 . "%A1EFE4")
     (300 . "%299BA6")
     (320 . "%2896B5")
     (340 . "%2790C3")
     (360 . "%66D9EF"))))
 '(vc-annotate-very-old-color nil)
 '(weechat-color-list
   (unspecified "%272822" "%20240E" "%F70057" "%F92672" "%86C30D" "%A6E22E" "%BEB244" "%E6DB74" "%40CAE4" "%66D9EF" "%FB35EA" "%FD5FF0" "%74DBCD" "%A1EFE4" "%F8F8F2" "%F8F8F0")))
(add-to-list 'package-archives '("gnu" . "http://elpa.gnu.org/packages"))
(add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/"))
(add-to-list 'package-archives '("marmalade" . "https://marmalade-repo.org/packages/"))
(add-to-list 'package-archives '("melpa-stable" . "http://stable.melpa.org/packages/"))
(add-to-list 'custom-theme-load-path "~/.emacs.d/themes/")
(load-theme 'monokai t)
(require 'auto-complete-clang)

;; Set keyboard shortcuts
(cua-mode t)
;; Meta shortcuts
(global-set-key (kbd "M-a") 'compile)
(global-set-key (kbd "M-g") 'gdb)
(global-set-key (kbd "M-r") 'package-refresh-contents)
(global-set-key (kbd "M-s") 'ispell-buffer)
(global-set-key (kbd "M-t") 'term)
(global-set-key (kbd "M-w") 'browse-web)
;; Ctrl shortcuts
(global-set-key (kbd "C-a") 'mark-whole-buffer)
;(global-set-key (kbd "C-c") 'kill-ring-save)
(global-set-key (kbd "C-e") 'elfeed)
(global-set-key (kbd "C-f") 'isearch-forward)
(global-set-key (kbd "C-l") 'load-file)
(global-set-key (kbd "C-o") 'menu-find-file-existing)
(global-set-key (kbd "C-p") 'package-install)
(global-set-key (kbd "C-s") 'save-buffer)
;(global-set-key (kbd "C-v") 'yank)
;(global-set-key (kbd "C-x") 'kill-region)
(global-set-key (kbd "C-z") 'undo)
;; Ctrl-Shift shortcuts
(global-set-key (kbd "C-S-o") 'dired)
(global-set-key (kbd "C-M-r") 'redraw-display)

;; Set environment variables
;(setq initial-buffer-choice "~/Documents/CodeLite/CPP-Math-Projects/Simple-Pendulum")
(setq ring-bell-function 'ignore)
(setq compilation-read-command nil)

;; if ~/.emacs.d/vendor/coffee-mode is absent git clone defunkt/coffee-mode to it!
(add-to-list 'load-path "~/.emacs.d/vendor/coffee-mode")
(add-to-list 'load-path "~/.emacs.d/elpa/company-lua-20161227.1139")
(require 'coffee-mode)
(require 'company)
(require 'company-lua)
(require 'company-tern)
(global-company-mode t)
(defun my-lua-mode-company-init ()
(add-to-list 'company-backends '((company-lua
                                 (company-tern))))
(add-hook 'lua-mode-hook #'my-lua-mode-company-init)

(add-to-list 'auto-mode-alist '("\\.coffee$" . coffee-mode))
(add-to-list 'auto-mode-alist '("Cakefile" . coffee-mode))
(progn
(autoload 'octave-mode "octave")
(setq auto-mode-alist (cons '("\\.m\\'" . octave-mode) auto-mode-alist))) 

(dolist (package '(auto-complete-clang auto-complete markdown-mode lua-mode gnuplot-mode elfeed dockerfile-mode))
 (unless (package-installed-p package)
   (package-install package))
   (require package))
(custom-set-faces
 ;; custom-set-faces was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 )

(speedbar 1)
(setq speedbar-directory-unshown-regexp "^\\(CVS\\|RCS\\|SCCS\\|\\.\\.*$\\)\\'")
(setq speedbar-directory-unshown-regexp "^$")
```

**Comment by ptrv** [comment]
Ok, I have double checked. It works for me.
When you are in in a lua file and put your curser after the dot for example

```lua
math.
```
can you post the output of the `company-diag`?

**Comment by fusion809** [comment]
Sorry? How do I get the output of company-diag when I have my cursor at the end of math.? Sorry if this falls under the category of a 'noob' question, I must confess Vim is my primary editor but I am interested in Emacs.

**Comment by ptrv** [comment]
Hit `M-x` then type `company-diag` and then hit `RET`
(`M-x` means Meta-x, which is usually the alt key)

**Comment by ptrv** [comment]
I tried the config you posted and also with it I make company-lua work.

I commented out some of the lines though, but nothing company-lua related, except I removed the company-tern completer.

I also removed auto-complete and auto-complete-clang because this is a different completion engine which should be used along side company-mode.

**Comment by ptrv** [comment]
This is the config I used:

```emacs-lisp
(require 'package)
(package-initialize)
(setq use-package-always-ensure t)
(add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/"))

(dolist (package '(lua-mode company-lua company))
  (unless (package-installed-p package)
    (package-refresh-contents)
    (package-install package)))

(require 'company)
(require 'company-lua)
(global-company-mode t)
(defun my-lua-mode-company-init ()
  (setq-local company-backends '(company-lua)))
(add-hook 'lua-mode-hook #'my-lua-mode-company-init)
```

**Comment by ptrv** [comment]
This is the output I get from `company-diag` when I run the command with the cursor after `math.p`

```
Emacs 27.0.50 (x86_64-pc-linux-gnu) of 2017-11-05 on linux-machine
Company 0.9.4

company-backends: (company-lua)

Used backend: company-lua
Major mode: lua-mode
Prefix: (#("p" 0 1
   (fontified t))
 . t)

Completions:
  #("pi" 0 2 (doc "The value of pi." returns "" args "" kind "value")) " [value]"
  #("pow" 0 3 (doc "Returns x^y.
(You can also use the expression x^y to compute this value.)" returns "(number)" args "(x: number, y: number)" kind "function")) "(x: number, y: number) -> (number) [function]"
```

**Comment by fusion809** [comment]
Well I'll be, with your ~/.emacs file for some reason it works perfectly. Completions are there. Must be other lines in my `~/.emacs` file that's messing it up. Well there are some things for which the completions are not there like when I type `io.` (hoping to end it with something like `read()`) no suggestions pop up. Likewise I use the lua-plot library and when I type `p:` after this bit of code I get no suggestions:

```lua
lp = require("lua-plot")
p = lp.plot{}
p:AddSeries(t, d2theta)
p:Show()
```

**Comment by ptrv** [comment]
Ok, so it's some configuration in your `~/.emacs` that interferes with company-lua.

Regarding completion for `io.`, there was indeed a bug. It should be fixed now with 29f6819de4d691e5fd0b62893a9f4fbc1c6fcb52 

This completer is very basic and does not support other libraries then standard library and the love2d game engine. The completion data is defined in files like in this one: https://github.com/ptrv/company-lua/blob/master/lua/api/baselib.lua
I haven't invested more time to do better completion and since I don't program in lua at the moment, it won't probably happen any time soon.

I'm open for Pull Requests for any improvements.
